File: app/src/main/java/top/easelink/lcg/ui/main/article/view/ArticleFragment.java
package top.easelink.lcg.ui.main.article.view;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.airbnb.lottie.LottieAnimationView;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import timber.log.Timber;
import top.easelink.framework.base.BaseFragment;
import top.easelink.lcg.BR;
import top.easelink.lcg.R;
import top.easelink.lcg.databinding.FragmentArticleBinding;
import top.easelink.lcg.mta.EventHelperKt;
import top.easelink.lcg.ui.ViewModelProviderFactory;
import top.easelink.lcg.ui.main.article.viewmodel.ArticleAdapter;
import top.easelink.lcg.ui.main.article.viewmodel.ArticleViewModel;
import top.easelink.lcg.ui.main.model.ReplyPostEvent;
import top.easelink.lcg.ui.webview.view.WebViewActivity;

import static top.easelink.lcg.mta.MTAConstantKt.EVENT_OPEN_ARTICLE;
import static top.easelink.lcg.ui.main.article.viewmodel.ArticleViewModel.FETCH_INIT;
import static top.easelink.lcg.utils.WebsiteConstant.SERVER_BASE_URL;


public class ArticleFragment extends BaseFragment<FragmentArticleBinding, ArticleViewModel>
        implements ArticleNavigator {

    @Inject
    ViewModelProviderFactory factory;
    private static final String KEY_URL = "KEY_URL";
    private LinearLayoutManager mLayoutManager;
    private String articleUrl;

    public static ArticleFragment newInstance(String url) {
        Bundle args = new Bundle();
        ArticleFragment fragment = new ArticleFragment();
        args.putString(KEY_URL, url);
        fragment.setArguments(args);
        EventHelperKt.sendEvent(EVENT_OPEN_ARTICLE);
        return fragment;
    }

    @Override
    public int getBindingVariable() {
        return BR.viewModel;
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_article;
    }

    @Override
    public ArticleViewModel getViewModel() {
        return ViewModelProviders.of(this, factory).get(ArticleViewModel.class);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EventBus.getDefault().register(this);
        getViewModel().setNavigator(this);
        Bundle args = getArguments();
        if (args != null) {
            articleUrl = args.getString(KEY_URL);
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);;
        setUp();
        setHasOptionsMenu(true);
        ((AppCompatActivity) getActivity()).setSupportActionBar(getViewDataBinding().articleToolbar);
        getViewModel().setUrl(articleUrl);
        getViewModel().fetchArticlePost(FETCH_INIT);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        EventBus.getDefault().unregister(this);
    }

    private void setUp() {
        mLayoutManager = new LinearLayoutManager(getContext());
        mLayoutManager.setOrientation(RecyclerView.VERTICAL);
        RecyclerView rv = getViewDataBinding().postRecyclerView;
        rv.setLayoutManager(mLayoutManager);
        rv.setItemAnimator(new DefaultItemAnimator());
        rv.setAdapter(new ArticleAdapter(getViewModel()));
        rv.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                int firstVisibleItemPosition = mLayoutManager.findFirstVisibleItemPosition();
                LottieAnimationView backToTop = getViewDataBinding().backToTop;
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    if (firstVisibleItemPosition == 0) {
                        backToTop.setVisibility(View.GONE);
                        backToTop.pauseAnimation();
                    } else {
                        backToTop.setVisibility(View.VISIBLE);
                    }
                } else if (newState == RecyclerView.SCROLL_STATE_DRAGGING) {
                    backToTop.setVisibility(View.GONE);
                    backToTop.pauseAnimation();
                }
            }
        });
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        menu.clear();
        inflater.inflate(R.menu.article, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_open_in_webview:
                WebViewActivity.startWebViewWith(SERVER_BASE_URL + articleUrl, getContext());
                break;
            case R.id.action_extract_urls:
                ArrayList<String> linkList = getViewModel().extractDownloadUrl();
                if (linkList != null && !linkList.isEmpty()) {
                    DownloadLinkDialog.newInstance(linkList).show(getFragmentManager());
                } else {
                    Toast.makeText(getContext(), R.string.download_link_not_found, Toast.LENGTH_SHORT).show();
                }
                break;
            case R.id.action_add_to_my_favorite:
                getViewModel().addToFavorite();
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void handleError(Throwable t) {
        Timber.e(t);
    }

    @Override
    public void showMessage(@StringRes int resId) {
        Toast.makeText(getContext(), resId, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void scrollToTop() {
        getViewDataBinding().postRecyclerView.smoothScrollToPosition(0);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(ReplyPostEvent event) {
        ReplyPostDialog.newInstance(event.getReplyUrl()).show(getChildFragmentManager());
    }
}

File: app/src/main/java/top/easelink/lcg/ui/main/article/view/ArticleFragment.kt
package top.easelink.lcg.ui.main.article.view

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.fragment_article.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import top.easelink.framework.topbase.ControllableFragment
import top.easelink.framework.topbase.TopFragment
import top.easelink.lcg.R
import top.easelink.lcg.ui.main.article.view.DownloadLinkDialog.Companion.newInstance
import top.easelink.lcg.ui.main.article.view.ReplyPostDialog.Companion.newInstance
import top.easelink.lcg.ui.main.article.view.ScreenCaptureDialog.Companion.TAG
import top.easelink.lcg.ui.main.article.viewmodel.ArticleAdapterListener.Companion.FETCH_POST_INIT
import top.easelink.lcg.ui.main.article.viewmodel.ArticleViewModel
import top.easelink.lcg.ui.main.model.ReplyPostEvent
import top.easelink.lcg.ui.main.model.ScreenCaptureEvent
import top.easelink.lcg.ui.main.source.model.Post
import top.easelink.lcg.ui.webview.view.WebViewActivity
import top.easelink.lcg.utils.WebsiteConstant
import top.easelink.lcg.utils.showMessage

class ArticleFragment: TopFragment(), ControllableFragment {

    companion object {
        fun newInstance(url: String): ArticleFragment {
            val args = Bundle().apply {
                putString(ARTICLE_URL, url)
            }
            val fragment = ArticleFragment()
            fragment.arguments = args
            return fragment
        }

        private const val ARTICLE_URL = "article_url"
        const val REPLY_POST_RESULT = 1000
    }

    private lateinit var viewModel: ArticleViewModel
    private var articleUrl: String = ""

    override fun isControllable(): Boolean {
        return true
    }

    override fun getBackStackTag(): String {
        return articleUrl
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        articleUrl = arguments?.getString(ARTICLE_URL).orEmpty()
        EventBus.getDefault().register(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_article, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this)[ArticleViewModel::class.java]
        initObserver()
        setUp()
        setupToolBar()
        viewModel.setUrl(articleUrl)
        viewModel.fetchArticlePost(FETCH_POST_INIT)
    }

    override fun onDetach() {
        super.onDetach()
        EventBus.getDefault().unregister(this)
    }

    private fun initObserver() {
        viewModel.articleTitle.observe(viewLifecycleOwner) {
            article_toolbar.title = it
        }
        viewModel.shouldDisplayPosts.observe(viewLifecycleOwner) {
            post_recycler_view.visibility = if (it) View.VISIBLE else View.GONE
        }
        viewModel.blockMessage.observe(viewLifecycleOwner) {
            block_text.text = it
            block_container.visibility = if (it.isNotEmpty()) View.VISIBLE else View.GONE
        }
        viewModel.isNotFound.observe(viewLifecycleOwner) {
            not_found_container.visibility = if (it) View.VISIBLE else View.GONE
        }
        viewModel.isLoading.observe(viewLifecycleOwner) {
            fetching_progress_bar.visibility = if (it) View.VISIBLE else View.GONE
        }
    }

    private fun setUp() {
        post_recycler_view.apply {
            val mLayoutManager = LinearLayoutManager(context).also {
                it.orientation = RecyclerView.VERTICAL
            }
            layoutManager = mLayoutManager
            itemAnimator = DefaultItemAnimator()
            adapter = ArticleAdapter(
                viewModel, this@ArticleFragment
            )

            viewModel.posts.observe(viewLifecycleOwner) {
                val url = it[0].replyUrl
                if (it.size > 0 && url != null) {
                    comment.apply {
                        visibility = View.VISIBLE
                    }.setOnClickListener {
                        val dialog = CommentArticleDialog.newInstance(url)
                        dialog.setTargetFragment(this@ArticleFragment, REPLY_POST_RESULT)
                        dialog.show(if (isAdded) parentFragmentManager else childFragmentManager)
                    }
                } else {
                    comment.visibility = View.GONE
                }
                (adapter as? ArticleAdapter)?.run {
                    clearItems()
                    addItems(it)
                }
            }
        }
    }

    private fun setupToolBar() {
        article_toolbar.apply {
            inflateMenu(R.menu.article)
            setOnMenuItemClickListener {
                when (it.itemId) {
                    R.id.action_open_in_webview -> WebViewActivity.startWebViewWith(
                        WebsiteConstant.SERVER_BASE_URL + articleUrl,
                        context
                    )
                    R.id.action_extract_urls -> {
                        viewModel.extractDownloadUrl()
                            ?.takeIf { list -> list.isNotEmpty() }
                            ?.run {
                                newInstance(this).show(if (isAdded) parentFragmentManager else childFragmentManager)
                            } ?: showMessage(R.string.download_link_not_found)
                    }
                    R.id.action_add_to_my_favorite -> viewModel.addToFavorite()
                    else -> {
                    }
                }
                return@setOnMenuItemClickListener true
            }
        }

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            REPLY_POST_RESULT -> {
                if (resultCode == 1) {
                    data?.getBundleExtra("post")
                        ?.getParcelable<Post>("post")
                        ?.let {
                            viewModel.addPostToTop(it)
                            post_recycler_view.scrollToPosition(1)
                        }
                    showMessage(R.string.reply_post_succeed)
                } else {
                    showMessage(R.string.reply_post_failed)
                }
            }
            else -> {
                super.onActivityResult(requestCode, resultCode, data)
            }
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: ReplyPostEvent) {
        newInstance(event.replyUrl, event.author).show(
            if (isAdded) parentFragmentManager else childFragmentManager
        )
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: ScreenCaptureEvent) {
        ScreenCaptureDialog
            .newInstance(event.imagePath)
            .show(
                fragmentManager = if (isAdded) parentFragmentManager else childFragmentManager,
                tag = TAG
            )
    }
}